### PR TITLE
[core] Make the BackgroundScheduler a singleton

### PR DESCRIPTION
**Pull request by tmpsantos**
- Do not carry it over everywhere as parameter, it is a shared
  instance anyway and the lifecycle is pretty much the app lifecycle
  from the moment we instantiate a map.
- Rename to BackgroundScheduler because it is a Scheduler that will
  do tasks in the background, we don't make assumptions if it is a
  thread pool or a single thread.
- Most importantly, remove the dependency from `core` on `platform`.

### CODE REVIEW

File: platform/default/src/mbgl/platform/thread_pool.cpp [hunk 1]
```diff
@@ -54,4 +76,9 @@ void ThreadPool::schedule(std::weak_ptr<Mailbox> mailbox) {
     cv.notify_one();
 }
 
+Scheduler& platform::GetBackgroundScheduler() {
+    static ThreadPool pool(4);
```
Comment (julianrex) [inline platform/default/src/mbgl/platform/thread_pool.cpp]: I think we should consider supporting some way for users of our SDKs to completely clean up after having used a mapview/our SDK, i.e. in this instance (and the similar one further in the PR) could we move the static to file scope, and then have some cleanup function that could deallocate the pool?

I'm thinking of situations where the mapview is just a secondary part of an application, which may be shown temporarily. Being able to completely clean up and have 0 lingering allocations would be nice.

File: platform/node/src/node_thread_pool.cpp [hunk 2]
```diff
@@ -35,3 +36,14 @@ void NodeThreadPool::Worker::WorkComplete() {
 }
 
 } // namespace node_mbgl
+
+namespace mbgl {
+namespace platform {
+
+Scheduler& GetBackgroundScheduler() {
+    static node_mbgl::NodeThreadPool pool;
+    return pool;
```
Comment (julianrex) [inline platform/node/src/node_thread_pool.cpp]: Similar here.

File: platform/default/src/mbgl/platform/thread_pool.cpp (hunk 1 above)
Comment (pozdnyakov) [reply platform/default/src/mbgl/platform/thread_pool.cpp]: Good point. Do we have already some `freeResorces()`-like API in our SDK?

**Comment by pozdnyakov** [comment]
> Do you have concerns about using `platform::GetBackgroundScheduler()` directly, rather than being injected? Seems that it more tightly couples those classes.

True, however IMO background worker is generic enough to be considered as a part of the project "foundation classes" :)

File: platform/default/src/mbgl/platform/thread_pool.cpp (hunk 1 above)
Comment (julianrex) [reply platform/default/src/mbgl/platform/thread_pool.cpp]: Not that I know of, @tmpsantos?

Comment (tmpsantos) [reply platform/default/src/mbgl/platform/thread_pool.cpp]: We have somethink on these lines on the `Renderer` class:
https://github.com/mapbox/mapbox-gl-native/blob/master/include/mbgl/renderer/renderer.hpp#L58

File: platform/default/src/mbgl/platform/thread_pool.cpp [hunk 3]
```diff
@@ -1,10 +1,32 @@
...
+#include <condition_variable>
+#include <mutex>
+#include <queue>
+#include <thread>
+
 namespace mbgl {
 
+class ThreadPool : public Scheduler {
+public:
+    ThreadPool(std::size_t count);
```
[the remark this answers is not in the document]
Comment (tmpsantos) [reply platform/default/src/mbgl/platform/thread_pool.cpp]: I miss `clang-tidy` :-)

File: platform/default/src/mbgl/platform/thread_pool.cpp (hunk 1 above)
Comment (tmpsantos) [reply platform/default/src/mbgl/platform/thread_pool.cpp]: > I'm thinking of situations where the mapview is just a secondary part of an application, which may be shown temporarily. Being able to completely clean up and have 0 lingering allocations would be nice.

That is a good point, I will address this concern.

File: src/mbgl/actor/scheduler.cpp [hunk 4]
```diff
@@ -1,19 +1,33 @@
...
 }
 
-Scheduler* Scheduler::GetCurrent() {
-    return current().get();
+// static
+std::shared_ptr<Scheduler> Scheduler::GetBackground() {
+    static std::weak_ptr<Scheduler> weak;
+    static std::mutex mtx;
+
+    std::lock_guard<std::mutex> lock(mtx);
```
Comment (alexshalamov) [inline src/mbgl/actor/scheduler.cpp]: why mutex? access to control block is thread safe.

**Pull request merged**